### PR TITLE
block access to child frames' contentWindow, contentDocument, place guard in blocking proxy to prevent an infinite loop

### LINKED ISSUE #11683: Fingerprinting protections bypassable
## Test plan

----

### Description
Many of the current fingerprinting protections rely on removing references to methods from the global object.  However, you can get alternate references to the same methods by inserting a configured iframe element and grabbing the references off `iframe.contentWindow` or `iframe.contentDocument`.

### Steps to Reproduce
```
//blocked
console.log(HTMLCanvasElement.prototype.toDataURL);

let iframe = document.createElement("iframe");
iframe.src = "https://www.brave.com/";
document.body.appendChild(iframe);

// Not blocked
console.log(iframe.contentWindow.HTMLCanvasElement.prototype.toDataURL);

// Do the bad stuff
let canvasToDataUrl = iframe.contentWindow.HTMLCanvasElement.prototype.toDataURL;
canvasToDataUrl.apply(someParentFrameCanvasElement);
```

**Actual result:**
You can get references to the fingerprinting related methods.

**Expected result:**
You shouldn't be able to get to these methods.

**Reproduces how often:** 100%

### Brave Version
All versions

**Reproducible on current live release:**
Yes

### LINKED ISSUE #11751: gmail load bar is broken by fingerprinting protection
see https://github.com/brave/browser-laptop/issues/11683#issuecomment-341224448

### PR DESCRIPTION
**Pull request by snyderp**
test plan for QA:

1. turn on 'block all fingerprinting' in about:preferences > Shields
2. login to gmail. it should work.
3. https://jsfiddle.net/3dhmjqco/1/ should show 1 fingerprinting method blocked

This PR is a second attempt to fix #11683. fixes #11751

It is identical to the [previous PR/attempt](https://github.com/brave/browser-laptop/pull/11708), with the exception that there is now a [loop guard](https://github.com/brave/browser-laptop/commit/8e294d6c15c4ee9436a41b0af0547a9c25d0f692#diff-be0dd6e3e16e84d0ea19bbd8e5380084R109) in place to prevent the kind of infinite loop the proxy block approach can get into in certain (rare, but not zero) call patterns.

Basically this approach is trying to thread the needle between minimizing site break-age from pulling expected stuff out (the proxy blocking approach) and falling back to the prior "null it out" approach when the call patterns don't allow that.

Bellow comments are copied over from prior effort to fix (they're all still accurate) …

> The fix is a little funky, since it requires injecting JS into child pages (the chrome.webFrame.setGlobal method wont work here, since we need to call to Object.definePropery) but it should be good.
> 
> Added a blocking access to fingerprinting methods on iframe.contentWindow test to test that things are actually getting blocked

### CODE REVIEW

**Comment by diracdeltas** [comment]
could you re-add a test case like the previous PR did?

**Comment by snyderp** [comment]
@diracdeltas Added the test back in.  I also removed some more semi-colons (hard thing to work out of muscle memory!)

**Comment by bsclifton** [comment]
@snyderp does this issue resolve https://github.com/brave/browser-laptop/issues/11751 ?

File: test/fixtures/fingerprinting-blocking-from-child-frames.html [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+<div id="target">
+proxy blocking being tested
+</div>
+<script>
+(function () {
+
+    let parentFrameMethod = window.HTMLCanvasElement.prototype.toDataURL
+
+    let iframe = document.createElement("iframe")
+    iframe.src = "https://www.brave.com/"
```
Comment (diracdeltas) [inline test/fixtures/fingerprinting-blocking-from-child-frames.html]: i think this fails now because brave.com sets the CSP frame-ancestors directive to prevent being iframed. you can use https://example.com instead.

**Comment by diracdeltas** [comment]
@snyderp i tried the manual test plan for the original issue (https://github.com/brave/browser-laptop/issues/11683#issuecomment-340532797) and it failed (https://jsfiddle.net/3dhmjqco/1/ shows 0 fingerprinting methods blocked instead of 1)

**Comment by snyderp** [comment]
@diracdeltas

3cc76d18879fe564d076d60888d64ff46b903cda has the URL for the test updated to https://example.com, and the test runs as expected again.

Regarding counting / detecting fingerprinting across iframes, I'm not sure this is possible (from an extension, if you were going to go deeper into chromium / C++ land you could).  The patch blocks all access to the DOMs of frames from parents, it can't / doesn't distinguish between fingerprinting and not fingerprinting behavior done from the parent, using the DOM of the child.

Before it was marking **all** cross frame DOM execution as fingerprinting related, and now it flags **none** of it as fingerprinting.  I'm not sure there is an option in between (again, w/o getting down into C++ land).

I can move back to the original "flag any cross frame DOM execution as fingerprinting related" option though if it'd be better

**Comment by snyderp** [comment]
@bsclifton Yep, it fixes #11751 !

**Comment by snyderp** [comment]
A third option here would be to replace the `contentWindow`/`contentDocument` properties with a similar but different modified proxy that triggered the "is fingerprinting" feedback whenever a method with the same name as one of the finger printing methods was called.

There would be some chance for a false positive here (e.g. the parent frame could be calling a different, user defined`toDataUrl` method, for example, unrelated to `HTMLCanvasElement.prototype.toDataURL`), but if getting some signal to the user is important (even if there is some chance of a false positive) then I could code something like that up.  

Its just a matter of whether Brave would prefer the risk of a false negative, or a false positive in this scenaro

**Comment by diracdeltas** [comment]
@snyderp yup, that's the approach i mentioned to you in slack. will have a PR in a bit for you to look at!

**Comment by diracdeltas** [comment]
@snyderp here's a follow-up commit to your work that cleans up the code a bit and reports iframe fingerprinting methods in the Brave shields panel: https://github.com/brave/browser-laptop/commit/f5190067604f499a1a0e33f43691640861806e94

however, I did something wrong because the infinite loop fix you added is not working on Gmail anymore :(

will look more tomorrow!

**Comment by snyderp** [comment]
@diracdeltas 

I dig the approach in f519006, but had a couple of thoughts:

1. You can still get to the fingerprinting methods through other paths in the DOM (e.x. `Object.getPrototypeOf(iframe.contentDocument.createElement("canvas")).toDataURL`,  `iframe.contentDocument.defaultView.HTMLCanvasElement.prototype.toDataURL`, etc).  Once you get any foothold into an unmodified DOM, its really difficult to prevent keypath lookups like this.  I'm unsure that you could comprehensibly block all access to functions via a proxy approach (why I erred on the side of "block everything in the child frame", though hopefully there is a better way!)

2. Small suggestion, but it might be better to grab a local reference to the `window.__braveReportBlock` and then deleting the global reference (to prevent the client page from seeing the reference)

3. I can add the loop guard back in if you'd like me too, just let me know :)

**Comment by diracdeltas** [comment]
@snyderp 

1. firstly, i forgot to stub out contentDocument - done in https://github.com/brave/browser-laptop/commit/e64bd0bd899fdcb9e8ec6060212da2855980e1b7. secondly, wouldn't a call like `iframe.contentDocument.defaultView.HTMLCanvasElement` return the HTMLCanvasElement of the parent document (which should be blocked already if canvas fp is working), since https://github.com/brave/browser-laptop/commit/e64bd0bd899fdcb9e8ec6060212da2855980e1b7#diff-be0dd6e3e16e84d0ea19bbd8e5380084R327 would proxy all calls on iframe.contentDocument to the `document` object? 

2. thanks, done in https://github.com/brave/browser-laptop/commit/bb72c40c6fe8a71977c22d6134b634e84d785965

3. that would be awesome!

**Comment by snyderp** [comment]
@diracdeltas 

Woah, this is a cool approach!  I didn't think about swapping out the child frame's refs for the parent frame's refs.  Thats neat!

I think it may cause other breaking though (for example, im not able to log into gmail with this patch). After debugging, the reason I found is that the parent frame wants to modify the child frame, but the parent frame ends up modifying itself accidentally.  Ex:

```
let iframe = document.createElement("iframe");
iframe.src = "https://www.example.com/";
document.body.appendChild(iframe);

// Divs is actually now the div elements in the parent frame, not the child
let divs = iframe.contentWindow.document.querySelectorAll("div");

// Divs from parent frame end up mangled, child frame is unaffected
divs.forEach(function (aDiv) {
    aDiv.parentNode.removeChild(aDiv);
});
```

If you log into gmail with bb72c40c6, you'll end up with a blank page for the above reason. (this gmail break is different than #11751, which was caused by an infinite loop)

Here is another option: replace the contentWindow/contentDocument properties with a new proxy object, that behaves exactly the same as the current blocking proxy, except it tigers the fingerprinting warning when any property with any of the finger printing methods is looked up: `getImageData`, `getLineDash`, `toDataURL`, etc.  There is still *some* chance of false positive, but the cases where a parent frame would want to benignly pull user defined methods out of a child frame seem so rare that I can't imagine it would be a problem.

If thats an appealing option, I can code it up this afternoon (and get the loop guard in that new, for-the-child-frames proxy too)

**Comment by snyderp** [comment]
The approach I mentioned above could also be modified to allow access to `contentWindow.postMessage` to, to allow that one specific, well advertised use for cross frame messaging

**Comment by snyderp** [comment]
@diracdeltas 

b4f9cac is another attempt at this.  It adds fingerprinting notification whenever anything in the iframe's contentWindow/contentDocument looks up anything with the names of any of the tracking related methods.  Its my best effort to bring together all the threads we've discussed so far

File: app/extensions/brave/content/scripts/blockCanvasFingerprinting.js [hunk 2]
```diff
@@ -267,4 +279,53 @@ if (chrome.contentSettings.canvasFingerprinting == 'block') {
     type: 'WebRTC',
     methodName: 'navigator.mediaDevices.enumerateDevices'
   })
+
+  var propertiesToReport = new Set(canvasMethods
+    .concat(canvasElementMethods)
+    .concat(webglMethods)
+    .concat(webglMethods)
```
Comment (diracdeltas) [inline app/extensions/brave/content/scripts/blockCanvasFingerprinting.js]: webgl is concat'ed twice

Comment (snyderp) [reply app/extensions/brave/content/scripts/blockCanvasFingerprinting.js]: Fixed in 

Sorry for all these issues, most just curious if the general approach is appealing for a patch.  If so, and if fadd50814 is not up to snuff, I'll give the PR a careful head-to-toe before sending it back to you

**Comment by diracdeltas** [comment]
it would be good to modify reportBlock('Iframe') so it returns the URL of the iframe or even just the top-level page instead of the URL of the content script:
screen shot 2017-11-15 at 2 05 23 am

**Comment by diracdeltas** [comment]
@snyderp thanks for looking at this again. i tried 1a079074ae73debd954932cc6b4472d9798c9af9 and still couldn't log in to gmail - did it work for you?

**Comment by snyderp** [comment]
@diracdeltas 
Hmm seems like something has changed (at the very least, the minified names of the gmail JS code have changed).  I'll keep looking into this, try to find something thats more robust.  My next idea is to create an iframe element before the first call to this.  If need be though, and there isn't a JS/extension level solution for this, would you accept a chromium level patch for this?

I'll try a few more revs in JS world, but feeling less confident at the moment that there is a "inject JS to modify the DOM" level solution that doesn't cause breaking too.

One option would be to take the Tor folks' approach, of patching these things in C++ land.  If thats of interest, i could work on those patches, though it might take a moment

Anyway, i'll spend another 24-48 hours brainstorming for some more JS level approaches here

**Comment by diracdeltas** [comment]
@snyderp definitely i think doing this at the C++ level is the right approach long-term, so it's a totally worthwhile investigation if you're up for it!

thanks again for looking into all these approaches

**Comment by snyderp** [comment]
@diracdeltas Ok, attempt bazillion, 8f2c19c1b74b121dacd9ab4c26647882e91e8ce3 should be the one.  It'll work with gmail, blocks the (i)frame contentWindow/contentDocument elements with a different-but-similar proxy, and reports the frame-based fingerprinting to the parent frame's URL.

I had to re-work the blocking code a bit, but I tried to document things well to keep things from getting squirly.  I didn't see a Brave document on docblock conventions, but if what i've added clashes with what you all use, just let me know and I can rework

**Comment by snyderp** [comment]
Not abandoning the C++ strategy, I agree that its the right way to go and prevents having to over-block in cases like this.  But, I hope in the short-to-medium term this approach is an improvement over whats there, until I can find time to do the deeper-dive

**Comment by diracdeltas** [comment]
w00t it works!! however i can't get the test to pass locally on MacOS (`npm run test -- --grep='blocking access to fingerprinting methods on iframe.contentWindow'`)

**Comment by snyderp** [comment]
Eggcellent!  5bec7186d44feb22fcbf0b1fc13e71741ef1f72e fixes the test.

I'm seeing the webgl blocking tests fail, (`blocks fingerprinting including WebGL on compact panel` and `blocks fingerprinting including WebGL`), but I see these failing on `master` too so I think this is unrelated

(if you're seeing those tests fail too, and also think they're unrelated to this issue, I'm happy to create new issues and PRs for those).

But otherwise, I hope / think this is all resolved

**Comment by diracdeltas** [comment]
test plan for QA:

1. turn on 'block all fingerprinting' in about:preferences > Shields
2. login to gmail. it should work.
3. https://jsfiddle.net/3dhmjqco/1/ should show 1 fingerprinting method blocked

**Comment by diracdeltas** [comment]
master: d8e0b3d3bd69ed228dcc576db720f20ab2efd6ee
0.21.x: 9ab65c489b4b651c478e121649c6c8e5bea4f5ed
0.20.x: be2e235711014b20888104440b71949909c8e4c0
0.19.x: d2675c32ba07b79a616871a795209c8eb0f8e24e

File: test/bravery-components/braveryPanelTest.js [hunk 3]
```diff
@@ -895,6 +895,20 @@ describe('Bravery Panel', function () {
...
+    it('blocking access to fingerprinting methods on iframe.contentWindow', function * () {
+      const url = Brave.server.url('fingerprinting-blocking-from-child-frames.html')
+      yield this.app.client
+        .tabByIndex(0)
+        .loadUrl(url)
+        .waitForUrl(url)
+        .openBraveMenu(braveMenu, braveryPanel)
+      yield changeFpSetting(this.app.client, blockFpOption)
+      yield this.app.client
+        .waitForTextValue(fpStat, '1')
```
Comment (luixxiul) [inline test/bravery-components/braveryPanelTest.js]: Should this be removed for Linux? The test looks to be failing due to this.

https://travis-ci.org/brave/browser-laptop/jobs/303759640#L5789

````
  3) Bravery Panel Adblock stats without iframe tests blocking access to fingerprinting methods on iframe.contentWindow:
     Promise was rejected with the following reason: timeout
  Error
````

Comment (diracdeltas) [reply test/bravery-components/braveryPanelTest.js]: i think this should pass on linux in general

**Comment by bsclifton** [comment]
master d56b737966deccbbe3debc52f608f9de8b44d3e9
0.21.x 9fd8f82bcf50e30fff80f43930d757b2e1ddbb29
0.20.x 9b3e84a14e9d47c05fa648a4a0764660fcbb44ae
0.19.x fe2fcedfdca94aeb29351fdd2388a3dca23fcedd
